### ISSUE TITLE
Mouse Follows Focus triggers on window close

### ISSUE THREAD

**Issue by chrishoage** [koekeishiya/kwm#172]
This is a very very minor annoyance, but wanted to ask about it anyway: 

From my understanding mouse-follows-focus should only be triggered when using focus giving commands (for example `window -f`). This all works as intended, however when closing a window, the next window to get focus triggers mouse follows focus and moves the mouse to the center of that window, even though I did not give a focus command (assuming the focus command came from kwm's internals to focus the next window upon close)

Is this expected / intended behavior? Or is Mouse Follows Focus only supposed to be triggered when a focus command is given? 

Thank you!

**Comment by koekeishiya** [comment]
>  (assuming the focus command came from kwm's internals to focus the next window upon close)

This is correct. When a window closes, Kwm will automatically give focus to the window that would fill up the space of the removed container.

> however when closing a window, the next window to get focus triggers mouse follows focus and moves the mouse to the center of that window,

This is intended behaviour, as I personally found it annoying to have the focus change on window close, without applying mouse-follows-focus.

**Comment by chrishoage** [comment]
> I personally found it annoying to have the focus change on window close, without applying mouse-follows-focus.

Got it, to me it makes more sense to only move the mouse when the user gives a focus command (user has intent to use whatever window they focused) where as when a user closes a window, they don't necessarily intend to use the next window in line. 

Thanks for answering my question, feel free to close this if there is no action to take.

**Comment by koekeishiya** [comment]
When I get to implement some sort of callback system, that would probably allow the user to override this behaviour. Still have to try and figure out the best approach for that, so no estimate for when that will happen for now.

**Comment by chrishoage** [comment]
This behavior is actually problematic now that `cmd-tab` and friends properly notify kwm about focus. If a app creates a dialog (like Chromes `alert` or some applications `update` dialog`) it creates it with out regard to where the origin window is. With `focus-follows-mouse` enabled moving to deal with this dialog causes the window behind it to bring into focus, `cmd+tab` back to the app that created the dialog snaps the mouse back to the center of the window. Rinse repeat. (very frustrating).

Ideally `kwmc config focus standby-on-float` and `kwmc config float-non-resizable` would work 100% of the time, but some windows slip though (like the types I mentioned) causing windows that are challenging to deal with. 

To mitigate this, I propose that `mouse-follows-focus` only is triggered  though `kwmc` commands, not other focus events (like `cmd+tab` or closing a window).

**Comment by koekeishiya** [comment]
I did some changes to this last nigh, does it work better now?

**Comment by chrishoage** [comment]
I made this to illustrate my problem https://www.dropbox.com/s/vegvpxwakkshndy/cmdtab.mov?dl=0

In the 2nd half of the movie I am cmd+tab to try and get my mouse over the alert dialog, and can't. 1st half of the movie shows how focus follows mouse hides the dialog (expected behavior)

**Comment by koekeishiya** [comment]
> In the 2nd half of the movie I am cmd+tab to try and get my mouse over the alert dialog, and can't.

That makes sense, Kwm does not try to perform mouse-follows-focus on cmd+tab. I assume this is something you want, and I think it seems reasonable. 

If I'm not mistaken, Kwm should now only do mouse-follllows-focus upon `window -f ..` , possibly `screen -f ..` and upon AddWindowToBSPTree

These popup windows annoy me quite a bit and I wish I could have Kwm automatically turn all of them to "topmost windows". I have not found any way to do this for windows in OSX without having to perform code-injection into the dock (which also requires the new security feature of El Capitan to be disabled).

**Comment by chrishoage** [comment]
> That makes sense, Kwm does not try to perform mouse-follows-focus on cmd+tab. I assume this is something you want, and I think it seems reasonable.

Apologies, I'm not sure I completely follow. Did you mean to say "does try" vs "does not try", because `mouse-follows-focus` on `cmd-tab` is (or at least appears) to cause the issue in the video. 

> If I'm not mistaken, Kwm should now only do mouse-follllows-focus upon window -f .. , possibly screen -f .. and upon AddWindowToBSPTree

Yeah, those seem like the only cases (to me) where `mouse-follows` should trigger. Other focus changing operations (`cmd-tab` etc) should not. 

> These popup windows annoy me quite a bit and I wish I could have Kwm automatically turn all of them to "topmost windows". I have not found any way to do this for windows in OSX without having to perform code-injection into the dock (which also requires the new security feature of El Capitan to be disabled).

Yeah, in a perfect world, they would be picked up by `kwmc config focus standby-on-float` and `kwmc config float-non-resizable` and then focus follows would just be turned off and the problem in the video would go away.

**Comment by koekeishiya** [comment]
> Apologies, I'm not sure I completely follow. Did you mean to say "does try" vs "does not try", because mouse-follows-focus on cmd-tab is (or at least appears) to cause the issue in the video. 

Actually I think I may be misunderstanding the problem.

When you cmd+tab to a different application, not a different window of the same application,  Kwm does perform mouse-follows-focus.

When you cmd+` to switch window withing an application, Kwm does not perform mouse-follows-focus.

In addition to this, the only places where mouse-follows-focus happens is
> If I'm not mistaken, Kwm should now only do mouse-follllows-focus upon window -f .. , possibly screen -f .. and upon AddWindowToBSPTree

If you would expect/prefer it to work differently, just write it out and I'll fix it during the evening.

**Comment by chrishoage** [comment]
> When you cmd+tab to a different application, not a different window of the same application, Kwm does perform mouse-follows-focus.

That is what is happening in the video. Due to `focus-follows-mouse` and the location of the dialog (which can't be controlled) I am required to `cmd-tab` between the application focused by `focus-follows` and the application that created the dialog. 

I place my mouse over the location where the dialog is, and `cmd-tab` back to the application that created the dialog. Doing this triggers `mouse-follows` moving my mouse, therefore not allowing me to deal with the dialog. 

>  If you would expect/prefer it to work differently, just write it out and I'll fix it during the evening.

Basically what you said: "only do mouse-follllows-focus upon window -f .. , possibly screen -f .. and upon AddWindowToBSPTree"

Right now `cmd-tab` also triggers it, I request that `cmd-tab` does not trigger it until I time where these dialogs can be handled by kwm more elegantly. (perhaps  by turning off `focus-follows` with `kwmc config focus standby-on-float` and `kwmc config float-non-resizable` until the dialog has been dealt with)

**Comment by koekeishiya** [comment]
Aight thanks for clarifying. Try the latest commit and see if it is usable.

**Comment by chrishoage** [comment]
Indeed, that works! thank you!
